sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,362 @@
+                w = coef
+            raw_prediction = X @ w + intercept
+        else:
+            # reshape to (n_classes, n_dof)
+            w = coef.reshape(self._loss.n_classes, -1)
+            if self.fit_intercept:
+                intercept = w[:, -1]
+                w = w[:, :-1]
+            else:
+                intercept = 0.0
+            raw_prediction = X @ w.T + intercept
+
+        return w, intercept, raw_prediction
+
+    def loss(self, coef, X, y, sample_weight=None, alpha=0.0, n_threads=1):
+        """Compute the loss as sum over point-wise losses.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+        y : C/F-contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+        """
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss = self._loss.loss(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if w.ndim == 1:
+            return loss + 0.5 * alpha * (w @ w)
+        else:
+            return loss + 0.5 * alpha * squared_norm(w)
+
+    def loss_gradient(self, coef, X, y, sample_weight=None, alpha=0.0, n_threads=1):
+        """Computes the sum/average of loss and gradient.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+        y : C/F-contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+
+        gradient : ndarray of shape (n_dof,) or (n_classes * n_dof)
+             The gradient of the loss as ravelled array.
+        """
+        n_features, n_classes = X.shape[1], self._loss.n_classes
+        n_dof = n_features + self.fit_intercept
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss, gradient = self._loss.loss_gradient(
```

Comment:
To spare the creation of arrays for gradient and hessian in each call to `loss_gradient` and the like, a follow-up PR could initialize them before calling an optimizer and then reuse them.